### DISCUSSION TITLE
Incremental migration off of SqlAlchemy-stubs possible?

### DISCUSSION THREAD

**Discussion by ghost** [sqlalchemy/sqlalchemy#11235] [Usage Questions]
According to the migration documentation:

> End-user code that passes all checks under SQLAlchemy 1.4 with the Mypy plugin may incrementally migrate to the new structures, once that code is running exclusively on SQLAlchemy 2.0. See the section [ORM Declarative Models](https://docs.sqlalchemy.org/en/20/changelog/whatsnew_20.html#whatsnew-20-orm-declarative-typing) for background on how this migration may proceed.

I have a large project that is running SQLAlchemy 2.0 but still has the SQLAlchemy-stubs package installed, and the mypy plugin `sqlmypy`. I want to remove the stubs and the plugin but am having trouble doing so in an incremental fashion, even though the docs indicate it is possible.

For example, if I take one db file and switch out `Column` to `mapped_column` and use the import
`from sqlalchemy.orm import mapped_column`, I will get the error: `Module "sqlalchemy.orm" has no attribute "mapped_column"`

If I fully delete the stubs package and use the new types, I have 30,000 type errors I need to fix, which is too many for a single MR.

Any suggestions? Am I missing an easy way to achieve this? I tried adding mypy stubs for `mapped_column` and added a `py.typed` file that contained `partial` in it, but the partial aspect doesn't seem to work. I also tried manually re-exporting everything from the stubs `__init__.pyi` file, but that also didn't seem to work. I would get errors like this:

```
mypy_stubs/sqlalchemy/__init__.pyi:4: error: Module "sqlalchemy" has no attribute "Index"  [attr-defined]
mypy_stubs/sqlalchemy/__init__.pyi:4: error: Module "sqlalchemy" has no attribute "Boolean"  [attr-defined]
mypy_stubs/sqlalchemy/__init__.pyi:4: error: Module "sqlalchemy" has no attribute "Integer"  [attr-defined]
...
```

While my `mypy_stubs/sqlalchemy/__init__.pyi` file looked like this:
```
from sqlalchemy import Index as Index, Boolean as Boolean, Integer as Integer, DateTime as DateTime, text as text, Text as Text, Column as Column, JSON as JSON, String as String
```

I'm just not sure how to have one file use types from the stubbed package and one to use the legit types from the sqlalchemy 2.0 package.

**Comment by zzzeek** [comment]
It sounds like you are using the Dropbox-maintained sqlalchemy-stubs package.   We have nothing to do with that package and instead we offer [sqlalchemy2-stubs](https://pypi.org/project/sqlalchemy2-stubs/) which has a different mypy plugin.  I think you'd want to get that package installed and working first and that's the mypy plugin that's referred to in the above mentioned documentation.

**Comment by ghost** [comment]
yea I think there was some confusion when I tried to use sqlalchemy2-stubs a few months ago. In [this comment thread](https://github.com/sqlalchemy/sqlalchemy/discussions/10949#discussioncomment-8325160) I said I was unable to upgrade to  `sqlalchemy2-stubs` because of [this issue](https://github.com/sqlalchemy/sqlalchemy2-stubs/issues/264). The response was "oh....the old dropbox stubs....yah we dont have any compatibility with those at all, sorry".  Since I was asking for help with `sqlalchemy2-stubs`, I thought that person mean _those_ types were the unsupported dropbox stubs and not supported at all.

I now see that there was some miscommunication. However I'm back to the original problem. I'll start down that path. Looks like there were some new comments let on the issue I encountered. Maybe I can work around it somehow.

**Comment by ghost** [comment]
Actually looks like I'm out of luck - I'd have to actually downgrade to SqlAlchemy 1.4, since I hit this issue now:

> mypy.ini:8: error: Error importing plugin "sqlalchemy.ext.mypy.plugin": The SQLAlchemy mypy plugin in SQLAlchemy 2.0 does not work with sqlalchemy-stubs or sqlalchemy2-stubs installed, as well as with any other third party SQLAlchemy stubs. Please uninstall all SQLAlchemy stubs packages. [misc]

😭

**Comment by zzzeek** [comment]
moving this back to #10949
